### PR TITLE
feat: LSDV-3896-1: Pannel tabs functionality

### PR DESCRIPTION
**Pull request by Travis1282**
### PR fulfills these requirements
- [X] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [X] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [X] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [X] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
_(check all that apply)_
- [X] Product design
- [X] Frontend

### Describe the reason for change
New Feature for tabs in panels, draggable across windows

#### What does this fix?
_(if this is a bug fix)_

#### What is the new behavior?
tabs can be pulled out of panels and rearranged or collapsed 

#### What is the current behavior?
currently there are no tabs on panels 

#### What libraries were added/updated?
N/A

#### Does this change affect performance?
N/A

#### Does this change affect security?
N/A

#### What alternative approaches were there?
_(briefly list any if applicable)_

#### What feature flags were used to cover this change?
fflag_feat_front_dev_3873_labeling_ui_improvements_short

### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [X] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [X] unit

### Which logical domain(s) does this change affect?
_(for bug fixes/features, be as precise as possible. ex. Authentication, Annotation History, Review Stream etc.)_

### CODE REVIEW

File: src/components/AnnotationsCarousel/AnnotationButton.tsx
Comment (bmartel) [inline src/components/AnnotationsCarousel/AnnotationButton.tsx:1]: Something about these changes may be causing a regression from #1096. Just a suspicion it is this change, but it definitely is something related to annotation button functionality. 

https://user-images.githubusercontent.com/2242258/225401784-49b795f3-12e1-4a54-8453-1a788796f761.mov

Comment (Travis1282) [reply src/components/AnnotationsCarousel/AnnotationButton.tsx:1]: can you check again? Made updates

File: src/components/SidePanels/TabPanels/utils.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    lastCallTime[name] = now;
+    callback();
+    clearTimeout(timeouts[name]!);
+    timeouts[name] = setTimeout(() => {
+      lastCallTime[name] = 0;
+      timeouts[name] = null;
+      callback2();
+    }, delay * 2);
+  }
+};
```
Comment (bmartel) [inline src/components/SidePanels/TabPanels/utils.ts]: Can this be replaced with just using lodash debounce? Not sure what this does exactly as it appears to not be throttling but debouncing the calls.

File: src/components/SidePanels/TabPanels/Tabs.tsx [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  }, [tabRef]);
+
+
+  const handleMouseOver = (event: React.MouseEvent<HTMLElement>) => {
+    if (event.buttons === 1 || event.buttons === 3) {
+      if (dragging.current) return;
+      const isDropArea = determineDroppableArea(event.target as HTMLElement);
+
+      if (isDropArea) setDragOverSide(determineLeftOrRight(event));
+      setTimeout(()=> setDragOverSide(undefined), 200);
```
Comment (bmartel) [inline src/components/SidePanels/TabPanels/Tabs.tsx:86]: What is the significance of `200` as a timeout here?

File: src/components/SidePanels/TabPanels/Tabs.tsx [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+const Tab = (props: TabProps) => {
+  const { rootRef, tabTitle: tabText, tabIndex, panelKey, viewLength, children, active, panelWidth, transferTab, createNewPanel, setActiveTab } = props;
+  const [dragOverSide, setDragOverSide] = useState<DroppableSide | undefined>();
+  const tabRef = useRef<HTMLDivElement>();
+  const ghostTabRef = useRef<HTMLDivElement>();
+  const dragging = useRef(false);
+  const location = useRef({ panelKey, tabIndex });
+
+  useEffect(() => { location.current = { panelKey, tabIndex }; }, [panelKey, tabIndex]);
```
Comment (bmartel) [inline src/components/SidePanels/TabPanels/Tabs.tsx:18]: Suggesting the same here, we don't need the useEffect to assign a ref for purposes of keeping a non stale reference to values in event handlers/callback closures.

```suggestion
  location.current = { panelKey, tabIndex };
```

File: src/components/SidePanels/TabPanels/Tabs.tsx [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        const droppedOnIndices = dropTargetId.split('_');
+        const receivingPanel = droppedOnIndices[0];
+        const receivingTab = parseInt(droppedOnIndices[1]);
+        const dropSide = determineLeftOrRight(event, dropTarget as HTMLElement);
+
+        if ((tabIndex === receivingTab || receivingTab === viewLength) && panelKey === receivingPanel) return; 
+        dropSide && transferTab(tabIndex, panelKey, receivingPanel, receivingTab, dropSide);
+      }
+    },
+  }, [tabRef]);
```
Comment (bmartel) [inline src/components/SidePanels/TabPanels/Tabs.tsx:77]: Is this a dependencies array? References are stable so if it is a dependencies array we would want to probably put nothing here.

File: src/components/SidePanels/TabPanels/SideTabsPanels.tsx [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+  const commonProps: CommonProps = useMemo(() => {
+    return {
+      ...eventHandlers,
+      root: rootRef,
+      regions,
+      selection: regions.selection,
+      currentEntity,
+    };
+  }, [eventHandlers, rootRef, regions, regions.selection, currentEntity]);
```
Comment (bmartel) [inline src/components/SidePanels/TabPanels/SideTabsPanels.tsx:270]: same thing here, rootRef is a stable reference, it won't trigger dependency arrays.

File: src/components/SidePanels/TabPanels/PanelTabsBase.tsx [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        handlers.current.onPositionChange?.(key, nY, nX, true);
+      }
+    },
+
+    onMouseUp() {
+      const { current: key } = keyRef;
+
+      handlers.current.onSnap?.(key);
+    },
+  }, [headerRef, detached, visible, locked]);
```
Comment (bmartel) [inline src/components/SidePanels/TabPanels/PanelTabsBase.tsx:189]: Another one with stable refs in a dependency array. `headerRef`

**Review by bmartel — CHANGES_REQUESTED** [review]

File: src/components/SidePanels/TabPanels/PanelTabsBase.tsx [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+  panelViews,
+}) => {
+  const headerRef = useRef<HTMLDivElement>();
+  const panelRef = useRef<HTMLDivElement>();
+  const resizerRef = useRef<HTMLDivElement>();
+  const handlers = useRef({ onResize, onResizeStart, onResizeEnd, onPositionChange, onPositionChangeBegin, onVisibilityChange, onSnap });
+  const [resizing, setResizing] = useState<string | undefined>();
+  const keyRef = useRef(key);
+
+  useEffect(() => { keyRef.current = key; }, [key]);
```
[the remark this answers is not in the document]
Comment (Travis1282) [reply src/components/SidePanels/TabPanels/PanelTabsBase.tsx:64]: much better thanks!

File: src/components/SidePanels/SidePanels.tsx [hunk 8]
```diff
@@ -155,7 +155,7 @@ const SidePanelsComponent: FC<SidePanelsProps> = ({
   }, [updatePanel]);
 
   const spaceFree = useCallback((alignment: 'left' | 'right') => {
-    return Object.values(panelData).find(p => p.alignment === alignment && !p.detached) === undefined;
+    return isFF(FF_DEV_3873) ? true : Object.values(panelData).find(p => p.alignment === alignment && !p.detached) === undefined;
```
Comment (yyassi-heartex) [inline src/components/SidePanels/SidePanels.tsx:158]:
```suggestion
    return isFF(FF_DEV_3873) || Object.values(panelData).find(p => p.alignment === alignment && !p.detached) === undefined;
```

**Review by bmartel — APPROVED** [review]

**Pull request closed without merging**